docs: update contributing.md and package-generator (#840)
artalar/reatom: docs/src/content/docs/contributing.md

description: How to contribute to Reatom
---

> **Note:** we prefer English language for all communication.

## Creating an issue

Before creating an issue please ensure that the problem is not [already reported](https://github.com/artalar/reatom/issues).

If you want to report a bug, create a reproduction using StackBlitz or CodeSandbox. If you want to request a feature, add motivation section and some usage examples.

## Sending a Pull Request

1. fork and clone the repository
2. create a development branch from `v3`
3. install dependencies from the root of the repo (`node@18` and `npm@8` are required):
   ```sh
   npm install
   ```

>    Dependencies are installed for all packages, but only packages included in `@reatom/framework` (`core`, `utils`, `async` etc.) are built.

Reviewer: I don't think that declarative sentence works here.
```suggestion
   Note: this command installs dependencies for all packages, but only builds packages included in `@reatom/framework`, like `core`, `utils`, `async`, etc.
```

6. upload feature branch and create a [Pull Request](https://github.com/artalar/reatom/compare) to merge changes to `v3`
7. link your PR to the issue using a [closing keyword](https://help.github.com/en/articles/closing-issues-using-keywords) or provide changes description with motivation and explanation in the comment (example: `fix #74`)
8. wait until a team member responds

## Creating a package

The goal of Reatom ecosystem is to provide adapters for Web APIs and popular npm modules. Therefore, the process of creating a new package is almost identical to editing an existing one ([Sending a Pull Request](#sending-a-pull-request)), but you should also create the package using an interactive script ran in the repository root:

```sh
npm run package-generator
```

Add needed dependencies by running `npm install` in your package's directory. If you're making an adapter for a particular npm library (like `@reatom/npm-react` for React), the library should be saved as peer: `npm install --save-peer <LIBRARY>`

<!-- ??? -->
<!-- To add dependencies, add them manually to the `package.json` of the new package and install them from the root of the repo. -->

### Package naming rule

> Packages that integrate Reatom with external APIs should have their names prefixed with one of the following strings: `node-`, `npm-`, `web-`. For example, [`@reatom/npm-history`](https://reatom.dev/package/npm-history) provides an adapter for the [`history`](https://npmjs.com/history) package. Similiarly, a potential adapter for [Web History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API) would be named `@reatom/web-history`.

Reviewer: ```suggestion
Packages that integrate Reatom with external APIs should have their names prefixed with API or library source kind: `node-`, `npm-`, `web-`. For example, [`@reatom/npm-history`](https://reatom.dev/package/npm-history) provides an adapter for the [`history`](https://npmjs.com/history) package. Similarly, a potential adapter for [Web History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API) would be named `@reatom/web-history`.
```

## Coding guide

- bug fixes should also add tests that reproduce the addressed bug
- all new features should be tested and documented
<!-- - always use `@ts-expect-error` instead of `@ts-ignore` -->
- use `// @ts-ignore` if you not sure why error appears or you think it could be better, use `// @ts-expect-error` if you sure that error is a mistake <!-- ??? -->

## Commit messages

Commit messages should follow the [Conventional Commits](https://conventionalcommits.org) specification:

```
<type>[optional scope]: <description>
```

### Allowed `<type>`

- `chore`: any repository maintainance changes
- `feat`: code change that adds a new feature
- `fix`: bug fix
- `perf`: code change that improves performance
- `refactor`: code change that is neither a feature addition nor a bug fix nor a performance improvement
- `docs`: documentation only changes
- `ci`: a change made to CI configurations and scripts
- `style`: cosmetic code change
- `test`: change that only adds or corrects tests
- `revert`: change that reverts previous commits

### Allowed `<scope>`

> Name of directory of a package (`/packages/<scope>`)

Reviewer: ```suggestion
Package directory name. Eg: `/packages/effects` is scoped as `effects` 
```

### `<description>` rules

- should be written in English

> - should be in imperative form (like `change` instead `changed` or `changes`)

Reviewer: ```suggestion
- should be in imperative tense (like `change` instead `changed` or `changes`)
```

- should not be capitalized

> - should not have period at the end

Reviewer: ```suggestion
- should not have period (`.`) at the end
```

### Package naming rule

> Packages that integrate Reatom with external APIs should have their names prefixed with API or library source kind: `node-`, `npm-`, `web-`. For example, [`@reatom/npm-history`](https://reatom.dev/package/npm-history) provides an adapter for the [`history`](https://npmjs.com/history) package. Similarly, a potential adapter for [Web History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API) would be named `@reatom/web-history`.Name of directory of a package (`/packages/<scope>`)

Reviewer: ```suggestion
Packages that integrate Reatom with external APIs should have their names prefixed with API or library source kind: `node-`, `npm-`, `web-`. For example, [`@reatom/npm-history`](https://reatom.dev/package/npm-history) provides an adapter for the [`history`](https://npmjs.com/history) package. Similarly, a potential adapter for [Web History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API) would be named `@reatom/web-history`.
```
